### ISSUE TITLE
Some benchmarks 

### ISSUE THREAD

**Issue by dangershony** [MithrilMan/MithrilShards#4]
This was posted on discord discussions

![image](https://user-images.githubusercontent.com/7487930/75292087-6def9300-581b-11ea-9277-d511610af947.png)

![image](https://user-images.githubusercontent.com/7487930/75292144-86f84400-581b-11ea-8339-a37a32a46d4f.png)

![image](https://user-images.githubusercontent.com/7487930/75292170-97102380-581b-11ea-8c91-814dced181fd.png)

![image](https://user-images.githubusercontent.com/7487930/75292211-adb67a80-581b-11ea-82df-2add7e91dc16.png)

![image](https://user-images.githubusercontent.com/7487930/75292261-bdce5a00-581b-11ea-9477-34ef140e0253.png)

**Comment by MithrilMan** [comment]
![image](https://user-images.githubusercontent.com/5107375/75293968-95e4f380-5827-11ea-9280-90bd3fc0a451.png)

mithrilshard parallelized is 5x faster than nbitcoin current implementation and almost half allocation
(i'm looking at 2000 headers results, last 4 lines
SequentialBitcoin is how currently it performs hash generation (one by one in a for loop), I added ParallelizedNBitcoin to show the performance improvement using Parallel.Foreach
pretty straightforward
  ```
[Benchmark]
public object SequentialNBitcoin()
{
   foreach (var header in headersNBitcoin)
   {
      header.GetHash();
   }
   return null;
}

[Benchmark]
public object ParallelizedNBitcoin()
{
   return Parallel.ForEach(headersNBitcoin, header => header.GetHash());
}
```

**Comment by MithrilMan** [comment]
this is a proper doublesha comparison against NBitcoin
![image](https://user-images.githubusercontent.com/5107375/75294129-f07e4f80-5827-11ea-80c4-2587f033a37e.png)

**Comment by MithrilMan** [comment]
![image](https://user-images.githubusercontent.com/5107375/75294159-012ec580-5828-11ea-913f-8600403e8c5a.png)
this one is quite interesting, I was testing different way to serialize bytes into hex strings, I came up with the green ones, I knew already which to pick (the one with Span) but wanted anyway to have some benchmark, actually it would save one more allocation because for this test I passed a byte array in, but in my case I've already a readonly span so I save even more
the interesting thing here, is that this is a generic Hex Encoder that produce hex string that's needed, if I get right the protocol, to generate the block header hash, basically the block id (that's my next step)
and so will be used very much
as you see, I've better performance and even better allocation
something more to put in the bank :slight_smile:
let me push the benchmark if someone is interested
https://github.com/MithrilMan/MithrilShards/commit/9d9aacb79f3ca1d12190048dad742499bd694853
candidates that I'm going to merge in my codebase are ConvertSpan and ConvertSpanReverse (will be renamed of course)

**Comment by MithrilMan** [comment]
another about block locator
https://media.discordapp.net/attachments/664593513769336875/665402291691716609/unknown.png

the one named Log is my implementation, Loop is NBitcoin
It's not completed, this benchmark just computes the indexes to fetch, I'd expect even more improvement on allocation because I won't do the final ToArray (copy) when populating the BlockLocator
https://github.com/MithrilMan/MithrilShards/commit/4759f4ec392a0b2333a86d02a69c95760ed6f893
basically I compute ahead the number of sent hashes using log base 2
https://github.com/MithrilMan/MithrilShards/blob/4759f4ec392a0b2333a86d02a69c95760ed6f893/MithrilShards.P2P.Benchmark/Benchmarks/Spot/BlockLocatorBuilding.cs#L19-L38
and the adapt the loops to do the job

the Loop method instead is the one used by NBitcoin (that just ported the one stated in the protocol documentation https://en.bitcoin.it/wiki/Protocol_documentation#getblocks)

**Comment by MithrilMan** [comment]
![image](https://user-images.githubusercontent.com/5107375/84444128-4cd37680-ac41-11ea-86af-34c90c15aa63.png)

I'm very happy about my optimization on allocation (that's more important than CPU in our scenario)
and I improved bitcoin core shift operators code too, I could open an issue or PR about that, basically I spare some pass in <<= and >>=
to recap: all test passed, 1 /4 of NBitcoin memory allocation and more than 100% gain in speed, enough for Target type!

Actually the code is on a branch, this is the code that I ported from bitcoin core from shift operators and improved too

https://github.com/MithrilMan/MithrilShards/blob/ff81533d3a285117848d67aed1ae93edeb27e6cd/src/MithrilShards.Chain.Bitcoin/DataTypes/Target.Operators.cs#L35-L85
